### ISSUE TITLE
Add regularization term to AMD?

### ISSUE THREAD

**Issue by ali-ramadhan** [climate-machine/Oceananigans.jl#335]
Right now AMD blows up if any of the fields do not have noise added to them due to dividing by velocity or tracer gradients.

@sandreza suggested that a small regularization term may be added to the denominator so that zero gradients do not produce NaNs.

**Comment by glwagner** [comment]
The issue I have with this solution is that there is no default value we can reasonably choose for such a parameter.

So, we can add the possibility for users to set a regularization parameter which is 0 by default. But by the time users realize that they need to set a regularization parameter, they could have already added noise to their initial condition. In essence, you'll have to add small numbers somewhere --- and I think its both easier and preferable (from the point of view of the correctness of the solution) to add these numbers to the initial condition rather than to change the model itself by adding small numbers in certain places. The notion of 'small' is better defined for the initial condition than, eg, for the strain rate (the denominator of the AMD fraction). 

Regardless of whether this is a good idea, I don't think we should consider changing the LES model until we have done some basic verification and understand its behavior.

**Comment by glwagner** [comment]
@ali-ramadhan, @sandreza, one solution is to add an `ifelse` statement to AMD that returns a zero nonlinear transport coefficient (viscosity or diffusivity) if the denominator is zero. I think this is correct behavior, because the denominator of the AMD transport coefficient being zero also implies that the flux divergence is zero (in principle --- this is only strictly true in the case that the gradients and transport coefficients are co-located, which is not how our algorithm works right now). Either way, it is reasonably correct to return a value of zero for the transport coefficient (unlike the addition of a regularization parameter, which we can show is definitely incorrect in some cases, hence my concern). 

In other words, the AMD viscosity function is

```julia
@inline function ν_ccc(i, j, k, grid::Grid{FT}, closure::AnisotropicMinimumDissipation, c,
                       eos, grav, u, v, w, T, S) where FT

    r = Δ²ₐ_uᵢₐ_uⱼₐ_Σᵢⱼ_ccc(i, j, k, grid, closure, u, v, w)
    ζ = Δ²ᵢ_wᵢ_bᵢ_ccc(i, j, k, grid, closure, eos, grav, w, T, S)
    q = tr_∇u_ccc(i, j, k, grid, u, v, w)

    νdagger = -closure.C * (r - ζ) / q 

    return max(zero(FT), νdagger) + closure.ν
end
```

I propose that we change the line with `νdagger` to

```julia
νdagger = ifelse(q != 0, -closure.C * (r - ζ) / q, zero(FT))
```

Let me know what you think.

**Comment by ali-ramadhan** [comment]
Ah interesting.

I'm wondering if this is an opportunity to short circuit AMD and speed it up a bit: If `q` is calculated first then you could `return zero(FT)` early and save some cycles.

**Comment by glwagner** [comment]
I'm fine with that too... so something like

```julia
if q == 0
    return zero(FT)
else
   ...
end
```

We were lead to believe that `ifelse` is faster on the GPU --- any concerns about that here? Presumably GPU cycles are cheap (but memory accesses are not as long as we aren't using shared memory...)

**Comment by ali-ramadhan** [comment]
Yeah the problem with `ifelse` is that it actually calculates both branches.

I thought it was faster on the GPU as well, but it depends on the use case. We've been using explicit `if` statements in the z-operators for a long time and it was still decently fast with millions of if statements per time step so I don't think using it in an AMD kernel will be bad.

We should still run `benchmark_turbulence_closures.jl` before and after of course.

**Comment by ali-ramadhan** [comment]
Should save some time as `Δ²ₐ_uᵢₐ_uⱼₐ_Σᵢⱼ_ccc` and `Δ²ᵢ_wᵢ_bᵢ_ccc` seem like expensive functions.

**Comment by glwagner** [comment]
This was implemented in #383.
